### PR TITLE
Representation of bstr as 1-byte int

### PR DESCRIPTION
**Pull request by gselander**
First draft to address #265

### CODE REVIEW

File: draft-ietf-lake-edhoc.md [hunk 1]
```diff
@@ -358,7 +358,8 @@ EDHOC includes the selection of connection identifiers (C_I, C_R) identifying a
 
 Connection identifiers may be used to correlate EDHOC messages and facilitate the retrieval of protocol state during EDHOC execution (see {{transport}}) or in subsequent applications of  EDHOC, e.g., in OSCORE (see {{ci-oscore}}). The connection identifiers do not have any cryptographic purpose in EDHOC.
 
-Connection identifiers in EDHOC are byte strings or integers, encoded in CBOR. One byte connection identifiers (the integers -24 to 23 and the empty CBOR byte string h'') are realistic in many scenarios as most constrained devices only have a few connections. An endpoint may choose to select only integer or only byte string connection identifiers.
+Connection identifiers in EDHOC are CBOR byte strings. Since most constrained devices only have a few connections short identifiers are desirable in many cases. However, except for the empty byte string h'' which encodes as as one byte (0x40), all byte strings are CBOR encoded as two or more bytes. Therefore EDHOC specifies certain byte strings to be represented as CBOR ints on the wire, see {{bstr-repr}}.
```
Comment (cabo) [inline draft-ietf-lake-edhoc.md:361]:
```suggestion
Connection identifiers in EDHOC are CBOR byte strings. Since most constrained devices only have a few connections, short identifiers are desirable in many cases. However, except for the empty byte string h'' which encodes as as one byte (0x40), all byte strings are CBOR encoded as two or more bytes. Therefore EDHOC specifies certain byte strings to be represented as CBOR ints on the wire, see {{bstr-repr}}.
``` 

(add comma for readability)

File: draft-ietf-lake-edhoc.md [hunk 2]
```diff
@@ -368,7 +369,29 @@ If connection identifiers are used by an application protocol for which EDHOC es
 
 ### Use of Connection Identifiers with OSCORE {#ci-oscore}
 
-For OSCORE, the choice of connection identifier results in the endpoint selecting its Recipient ID, see Section 3.1 of {{RFC8613}}, for which certain uniqueness requirements apply, see Section 3.3 of {{RFC8613}}. Therefore, the Initiator and the Responder MUST NOT select connection identifiers such that it results in same OSCORE Recipient ID. Since the Recipient ID is a byte string and an EDHOC connection identifier is either a CBOR byte string or a CBOR integer, care must be taken when selecting the connection identifiers and converting them to Recipient IDs. A mapping from EDHOC connection identifier to OSCORE Recipient ID is specified in {{edhoc-to-oscore}}.
+For OSCORE, the choice of connection identifier results in the endpoint selecting its Recipient ID, see Section 3.1 of {{RFC8613}}, for which certain uniqueness requirements apply, see Section 3.3 of {{RFC8613}}. Therefore, the Initiator and the Responder MUST NOT select connection identifiers such that it results in same OSCORE Recipient ID. Since the connection identifier is byte-valued, it is converted to an OSCORE Recipient ID equal to the byte string.
```
Comment (cabo) [inline draft-ietf-lake-edhoc.md:372]: "byte-valued"?  I thought it was a byte string.

File: draft-ietf-lake-edhoc.md [hunk 3]
```diff
@@ -476,7 +499,7 @@ Example: CWT or CCS can be identified by a key identifier using the 'kid' parame
...
+The COSE 'kid' parameter has byte string as value. To allow more one-byte key identifiers, which may be useful in many scenarios since constrained devices only have a few keys, we use the same integer representation as for connection identifiers, see {{bstr-repr}}. As stated in Section 3.1 of {{I-D.ietf-cose-rfc8152bis-struct}}, applications MUST NOT assume that 'kid' values are unique and several keys associated with a 'kid' may need to be checked before the correct one is found. Applications might use additional information such as 'kid context' or lower layers to determine which key to try first. Applications should strive to make ID_CRED_x as unique as possible, since the recipient may otherwise have to try several keys.
```
Comment (cabo) [inline draft-ietf-lake-edhoc.md:502]:
```suggestion
The value of a COSE 'kid' parameter is a byte string. To allow more one-byte key identifiers, which may be useful in many scenarios since constrained devices only have a few keys, we use the same integer representation as for connection identifiers, see {{bstr-repr}}. As stated in Section 3.1 of {{I-D.ietf-cose-rfc8152bis-struct}}, applications MUST NOT assume that 'kid' values are unique and several keys associated with a 'kid' may need to be checked before the correct one is found. Applications might use additional information such as 'kid context' or lower layers to determine which key to try first. Applications should strive to make ID_CRED_x as unique as possible, since the recipient may otherwise have to try several keys.
```

File: draft-ietf-lake-edhoc.md [hunk 4]
```diff
@@ -752,7 +775,7 @@ message_1 = (
   METHOD : int,
   SUITES_I : suites,
   G_X : bstr,
-  C_I : bstr / int,
+  C_I : bstr,
```
Comment (cabo) [inline draft-ietf-lake-edhoc.md:778]: Why this change?

File: draft-ietf-lake-edhoc.md [hunk 5]
```diff
@@ -805,7 +828,7 @@ message_2 SHALL be a CBOR Sequence (see {{CBOR}}) as defined below
 ~~~~~~~~~~~ CDDL
 message_2 = (
   G_Y_CIPHERTEXT_2 : bstr,
-  C_R : bstr / int,
+  C_R : bstr,
```
Comment (cabo) [inline draft-ietf-lake-edhoc.md:831]: Why this change?

**Review by cabo — COMMENTED** [review]
Here's a quick leaf-through.

**Comment by gselander** [comment]
@cabo Thanks. I actually thought about the text parts you commented on but I had worked too late to do something about it.

About the connection id type It should be 

C_x : bstr / -24..23,

or do you see any reason to keep bstr / int?

**Comment by cabo** [comment]
Maybe I would give `-24..23` a name... Not enough creativity today.  `obint` (one-byte int)?  Haven't checked that in all languages :-)
Yes, that narrow range is a more accurate representation for what you are saying in the text.

File: draft-ietf-lake-edhoc.md (hunk 2 above)
Comment (gselander) [reply draft-ietf-lake-edhoc.md]: Changed in 0ab48af

**Comment by gselander** [comment]
Right, let's first hear what people think.

File: draft-ietf-lake-edhoc.md (hunk 4 above)
Comment (marco-tiloca-sics) [inline draft-ietf-lake-edhoc.md:778]: Also consistently with the later editing in the "CDDL Definitions" appendix, this should be

`C_I : bstr / -24..23,`

File: draft-ietf-lake-edhoc.md (hunk 5 above)
Comment (marco-tiloca-sics) [inline draft-ietf-lake-edhoc.md:831]: Also consistently with the later editing in the "CDDL Definitions" appendix, this should be

`C_R : bstr / -24..23,`

File: draft-ietf-lake-edhoc.md [hunk 6]
```diff
@@ -368,7 +369,29 @@ If connection identifiers are used by an application protocol for which EDHOC es
...
+
+~~~~~~~~~~~
+Integer:               -24   -23  ...   -2    -1     0     1  ...   23
+CBOR encoding (1 byte): 37    36  ...   21    20    00    01  ...   17
+~~~~~~~~~~~
+{: #fig-int-one-byte title="One-Byte CBOR Encoded Integers"}
+{: artwork-align="center"}
+
+To allow more byte string identifiers that are only one byte on the wire, certain byte strings are defined to  have integer representations in the interval -24, ... ,23. These byte strings are those for which there is an integer in the interval -24, ..., 23 that happens to have this byte string as its CBOR encoding, see {{fig-bstr-as-int}} and compare {{fig-int-one-byte}}.
+
```
Comment (marco-tiloca-sics) [inline draft-ietf-lake-edhoc.md:388]: Is there a reason to limit this representation only to the interval -24..23?

If it rather considered all the integer values:

* There will be several more identifiers that can be transported as integers, again with 1 byte less overhead compared to the byte string representation. For example, the identifier 0x1839 can be transported as 0x1839 (CBOR integer 57) rather than 0x421839 (CBOR byte string).

* I believe it would be overall simpler for an implementation, that would simply have to check about integer vs. byte string, rather than specifically integer -24..23 vs. byte string.

The same applies for 'kid' values when 'kid' is used in ID_CRED_X.

Comment (gselander) [reply draft-ietf-lake-edhoc.md:388]: Without commenting on this proposal, I don't think we should optimise the sheer number of short identifiers.
There needs to be a balance here between what is easily specified, understood and coded. 

IMO, as long as we can have a number of 1-byte connection identifiers and key identifiers in EDHOC (to comply with the LAKE requirements) and support for the empty byte string in OSCORE (to reduce bytes in the most frequent message exchange) the main effort should be spent on making the EDHOC specification and the conversion between EDHOC and OSCORE identifiers as simple as possible. 

@marco-tiloca-sics Have you written down the procedure for reversing the representation, and for mapping back and forth between EDHOC and OSCORE?

One concern I have (with #267 in general) is with the pre-pending of connection identifiers done at transport layer. If we introduce special EDHOC representations, does the transport layer also need to know about that? (This was one reason for doing #274.)

File: draft-ietf-lake-edhoc.md (hunk 4 above)
Comment (gselander) [reply draft-ietf-lake-edhoc.md]: Overtaken by later proposals

File: draft-ietf-lake-edhoc.md (hunk 5 above)
Comment (gselander) [reply draft-ietf-lake-edhoc.md]: Overtaken by later proposals

File: draft-ietf-lake-edhoc.md (hunk 6 above)
Comment (gselander) [reply draft-ietf-lake-edhoc.md:386]: The proposal to apply the mapping for all integers is now included. I agree that the user of transport needs to provide the transport-ready representation, including prepended identifier, this is also noted in the update.

**Comment by gselander** [comment]
I think the PR is ready for review (thanks @cabo and @marco-tiloca-sics).

Short summary: we may not be able to register `kid` as bstr / int. This proposal is based on `kid` (and connection identifiers) intrinsically being byte strings, but can be represented as CBOR ints in EDHOC messages.

@chrysn @malishav @TimothyClaeys and others, please have a look.

**Comment by cabo** [comment]
I don't think enabling the shorter integer representation for all byte strings that parse as CBOR integers is worth it.

By staying with -24..23, the integerness can easily be tested, as in

b.size == 1 && !(b[0]&0xC0) && ((b[0] & 0x1f) < 0x18)

and the integer can be extracted as

(b[0] & 0x1f) ^ -((b[0] >> 4) & 0x1)

(untested code, but probably close).

Allowing all integers means that the CBOR library needs to be called and there needs to be a way in that to handle unsuccessful results.

All for a quite questionable additional benefit.

**Comment by cabo** [comment]
(Assuming the conversion to integer is mandatory -- if it is not, this becomes an interop testing nightmare.)

**Comment by gselander** [comment]
Thanks @cabo. The changes to any length of integer representation was made for perceived simplicity by @marco-tiloca-sics. And yes, we must clarify that this is mandatory.

**Comment by chrysn** [comment]
Let's get some numbers on what we get from just-bytes (0), -24..23 (1) vs. any-integer (2):

```
+---------+-+---------+-----------+-----------+
| size    | usable 0  | usable 1  | usable 2  |
+---------+-+---------+-----------+-----------+
| 1       | 1         | 49        | 49        |
| 2       | 256       | 208       | 720       |
| 3       | 64k       | 64k - 48  | ~3 x 64k  |
+---------+-+---------+-----------+-----------+
```

Clearly using any-integer is higher density (and I admit to having been the one who started going about these densities initially), but really where is it relevant? A system needs to use more than 257 identifiers before 1 vs 2 starts making a difference. And even then, things start to differ only for the EDHOC phase, there are (as there have always been) 256 1-byte and 64k 2-byte IDs for available for OSCORE, it's just that some of them (most of them, in either case) need an extra byte to inhabit CBOR encodings.

Constrained systems likely get away with picking their own identifiers exclusively from -24..23 anyway.

I think the larger question is "what is simpler?", and that depends by the use case:

* Picking a value from a series of pools (sorted by preference for OSCORE use and then preference for EDHOC use) unless it is taken:
  1. pools are h'', (-24..=23), h'00' to h'ff' with two holes, and the long strings.
  2. pools are h'', (-24..=23), h'00' to h'ff' with two holes, (-256..=-25), (24..256), h'0000' to h'ffff' with four holes etc.
* Encoding a picked value is trivial either way.
* Verifying that the peer encoded its ID correctly -- does the party receiving the C_x really need to do this?
* Decoding a peer's byte encoded C_x -- again, does this even ever need to be done? (The OSCORE contexts are best stored indexed by recipient ID)

At least if the latest two are answered "no" (but probably either way), I'd say that 1 is easier to implement, and while the coding density differs by a factor of ~3 for the 2-3 byte values, it's the same where it matters the most.

File: draft-ietf-lake-edhoc.md (hunk 6 above)
Comment (gselander) [reply draft-ietf-lake-edhoc.md:384]: Now reverted to one-byte integers only.

**Comment by gselander** [comment]
Sorry for the delay, now reverted to (1).

I didn't try to make the CDDL exclude the disallowed byte strings, but made a normative statement. 

@cabo @marco-tiloca-sics @chrysn Please check.

**Comment by marco-tiloca-sics** [comment]
It looks good to me.

**Comment by malishav** [comment]
The PR looks quite clear to me.

File: draft-ietf-lake-edhoc.md [hunk 7]
```diff
@@ -358,17 +358,46 @@ EDHOC includes the selection of connection identifiers (C_I, C_R) identifying a
...
 
 C_I and C_R are chosen by I and R, respectively. The Initiator selects C_I and sends it in message_1 for the Responder to use as a reference to the connection in communications with the Initiator. The Responder selects C_R and sends it in message_2 for the Initiator to use as a reference to the connection in communications with the Responder.
 
 If connection identifiers are used by an application protocol for which EDHOC establishes keys then the selected connection identifiers SHALL adhere to the requirements for that protocol, see {{ci-oscore}} for an example.
 
+### Representation of Byte String Identifiers {#bstr-repr}
+
+To allow identifiers with minimal overhead on the wire, certain byte strings are defined to have integer representations.
+
+The integers with one-byte CBOR encoding are -24, ..., 23, see {{fig-int-one-byte}}. This correspondence between integers and byte strings is a natural mapping between the byte strings with CBOR diagnostic notation h'00', h'01', ..., h'37' (except h'18' and h'19') and integers which are CBOR encoded as one byte.
```
Comment (emanjon) [inline draft-ietf-lake-edhoc.md:374]: > is a natural mapping between the byte strings with CBOR diagnostic notation h'00', h'01', ..., h'37' (except h'18' and h'19') and integers which are CBOR encoded as one byte.

There are 54 such byte strings and 48 integers which are CBOR encoded as one byte

File: draft-ietf-lake-edhoc.md [hunk 8]
```diff
@@ -476,7 +505,9 @@ Example: CWT or CCS can be identified by a key identifier using the 'kid' parame
 
 * ID_CRED_x = { 4 : key_id_x }, where key_id_x : kid, for x = I or R.
 
-Note that 'kid' is extended to support int values to allow more one-byte identifiers (see {{kid-header-param}} and {{kid-key-common-param}}) which may be useful in many scenarios since constrained devices only have a few keys. As stated in Section 3.1 of {{I-D.ietf-cose-rfc8152bis-struct}}, applications MUST NOT assume that 'kid' values are unique and several keys associated with a 'kid' may need to be checked before the correct one is found. Applications might use additional information such as 'kid context' or lower layers to determine which key to try first. Applications should strive to make ID_CRED_x as unique as possible, since the recipient may otherwise have to try several keys.
+The value of a COSE 'kid' parameter is a byte string. To allow one-byte encodings of ID_CRED_x with key identifiers 'kid', which is useful in scenarios with only a few keys, the integer representation of identifiers in {{bstr-repr}} is applied. For details, see {{asym-msg2-proc}} and {{asym-msg3-proc}}.
```
Comment (emanjon) [inline draft-ietf-lake-edhoc.md:508]: There should be a MUST here.

File: draft-ietf-lake-edhoc.md [hunk 9]
```diff
@@ -1122,7 +1153,9 @@ An implementation MAY support only Initiator or only Responder.
 
 An implementation MAY support only a single method. None of the methods are mandatory-to-implement.
 
-Implementations MUST support 'kid' parameters of type int. None of the other COSE header parameters are mandatory-to-implement.
+Implementations MUST support byte string identifiers represented as one-byte CBOR ints, as described in {{bstr-repr}}.
```
Comment (emanjon) [inline draft-ietf-lake-edhoc.md:1156]: > Implementations MUST support byte string identifiers represented as one-byte CBOR ints, as described in {{bstr-repr}}.

If we have MUST earlier in the document I think this sentence should be removed. "supported" gives the idea that this is not mandatory to encode as int, which would lead to a non-deterministic encoding.

File: draft-ietf-lake-edhoc.md [hunk 10]
```diff
@@ -1202,7 +1235,7 @@ An attacker observing network traffic may use connection identifiers sent in cle
...
 
-Information regarding the lengths of ID_CRED_I, ID_CRED_R, EAD_2, EAD_3, and EAD_4 are leaked to an passive attacker. Many COSE header parameters have variable length. To mitigate an attacker from differentiating endpoints with identifiers of different length, and implementation may e.g., only use fix length identifiers like 'kid' of length 1 for the Responders. Alternatively padding may be used to hide the length of e.g., certificates by value in 'x5chain' or 'c5c'.
+Information regarding the lengths of ID_CRED_I, ID_CRED_R, EAD_2, EAD_3, and EAD_4 are leaked to a passive attacker. Many COSE header parameters have variable length. To mitigate an attacker from differentiating endpoints with identifiers of different length, an implementation may, e.g., only use fix length identifiers like 'kid' with one-byte CBOR int representations for the Responders. Alternatively padding may be used to hide the length of, e.g., certificates by value in 'x5chain' or 'c5c'.
```
Comment (emanjon) [inline draft-ietf-lake-edhoc.md:1238]: > are leaked to a passive attacker.

are is too strong here

**Review by emanjon — APPROVED** [review]

File: draft-ietf-lake-edhoc.md (hunk 10 above)
Comment (gselander) [reply draft-ietf-lake-edhoc.md:1236]: Addressed by #282
